main/src/library/Nel.flix

```diff
@@ -220,6 +224,27 @@ namespace Nel {
         case x :: rs => f(x, i) :: mapWithIndexHelper(f, rs, i + 1)
     }
 
+    ///
+    /// Apply every function from `f` to every argument from `l` and return a non-empty list with all results.
+    /// For `f = f1, f2, ...` and `x = x1, x2, ...` the results appear in the order
+    /// `f1(x1), f1(x2), ..., f2(x1), f2(x2), ...`.
+    ///
+    pub def ap(f: Nel[a -> b & ef], l: Nel[a]) : Nel[b] & ef =  match f {
+        case Nel(f1, Nil) => map(f1, l)
+        case Nel(f1, rs)  => apHelperList(f1, rs, l, ks -> ks as & ef)
+    }
+
+    ///
+    /// Helper function for `ap`.
+    ///
+    def apHelperList(f: a -> b & ef, fs: List[a -> b & ef], l: Nel[a], k: Nel[b] -> Nel[b] & ef): Nel[b] & ef = match fs {
```

Comment:
Since we now have local defs (let-rec) the goal is to have all of these helpers inlined.

Reply:
Updated to use loop